agentops/partners/langgraph_callback_handler.py

```diff
@@ -0,0 +1,961 @@
+        self.agent_actions[str(run_id)].append(agent_event)
+        self.ao_client.record(agent_event)
+
+    @debug_print_function_params
+    def on_agent_finish(
+        self,
+        finish: AgentFinish,
+        *,
+        run_id: UUID,
+        parent_run_id: Optional[UUID] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Handle agent finish and ensure final output is logged."""
+        run_id_str = str(run_id)
+
+        agent_events = self.agent_actions.get(run_id_str)
+        if agent_events:
+            for agent_event in agent_events:
+                if agent_event.end_timestamp is None:
+                    agent_event.end_timestamp = get_ISO_time()
+                    self.ao_client.record(agent_event)
+            del self.agent_actions[run_id_str]
+        else:
+            logger.warning("Agent actions not found for run_id: %s", run_id)
+
+        action_event = ActionEvent(
+            action_type="agent_finish",
+            params={"log": finish.log},
+            returns=finish.return_values,
+            init_timestamp=get_ISO_time(),
+            end_timestamp=get_ISO_time()
+        )
+
+        if parent_run_id:
+            action_event.parent_id = str(parent_run_id)
+
+        self.ao_client.record(action_event)
+
+    @debug_print_function_params
+    def on_retry(
+        self,
+        retry_state: RetryCallState,
+        *,
+        run_id: UUID,
+        parent_run_id: Optional[UUID] = None,
+        **kwargs: Any,
+    ) -> Any:
+        action_event = ActionEvent(
+            params={
+                "attempt_number": retry_state.attempt_number,
+                "wait": retry_state.next_action.sleep,
+                "outcome": str(retry_state.outcome),
+            },
+            returns=str(retry_state),
+            action_type="retry",
+            init_timestamp=get_ISO_time(),
+        )
+        self.ao_client.record(action_event)
+
+    @debug_print_function_params
+    def on_text(
+        self,
+        text: str,
+        *,
+        run_id: UUID,
+        parent_run_id: Optional[UUID] = None,
+        **kwargs: Any,
+    ) -> Any:
+        text_event = ActionEvent(
+            params={"text": text, **kwargs},
+            action_type="text",
+            init_timestamp=get_ISO_time(),
+            end_timestamp=get_ISO_time(),
+        )
+        self.ao_client.record(text_event)
+
+    @property
+    def session_id(self):
+        raise DeprecationWarning(
+            "session_id is deprecated in favor of current_session_ids"
+        )
+
+    @property
+    def current_session_ids(self):
+        return self.ao_client.current_session_ids
+
+    def _cleanup_event(self, event_dict: Dict[str, Any], run_id: UUID) -> None:
+        """Safely remove an event from the event dictionary."""
+        run_id_str = str(run_id)
+        if run_id_str in event_dict:
+            del event_dict[run_id_str]
+
+    def _get_event_or_create_error(
```

Comment:
This function should be defined outside the class to work with both the `LanggraphCallbackHandler` and `AsyncLanggraphCallbackHandler`.